test/specs/flat-map.js

```diff
@@ -129,6 +129,23 @@ describe('flatMap', () => {
       expect(unsubs).to.equal(1)
     })
 
+    it('should not error when source ends in response to synchronous value', () => {
+      let didEnd = false
+      const src = stream()
+      src
+        .flatMap(x => Kefir.constant(x))
+        .onValue(x => {
+          if (x === 1) {
+            send(src, [end()])
+          }
+        })
+        .onEnd(() => {
+          didEnd = true
+        })
+      send(src, [value(1)])
+      expect(didEnd).to.be.true
```

Comment:
Instead of the `didEnd` setup, we should probably use `expect(src).to.emit([])` assert here. This will always verify what values are emitted by the `flatMap`'ed steam. My wonder / worry is the initial value never gets emitted.

I also don't think this test activates the stream, so it could pass without actually testing the code. You can look at `chai-kefir`'s assertions to help. Otherwise, I can take a quick look at this.

Reply:
Oh wait, `onValue` activates the stream... This mimics a circular dependency, so maybe this test is fine?

Reply:
I tried making the test like this:

```js
      const src = stream()
      const result = src
        .flatMap(x => Kefir.constant(x))
        .onValue(x => {
          if (x === 1) {
            send(src, [end()])
          }
        })
      expect(result).to.emit([end()], () => {
        send(src, [value(1)])
      })
```

but I didn't really like that this version of the test is also testing that the `value(1)` doesn't make it to the expect listener at the end. I don't mind that that happens, but I don't want the test to be dependent on that fact as it distracts from the real purpose of the test (and also if we ever change that behavior, I don't want this test to stand in the way). So using .onValue() and expect().to.emit() together doesn't seem to work out very well, and I don't think there's an unconvoluted way to accomplish this test with just expect().to.emit() instead of .onValue().

Reply:
> but I didn't really like that this version of the test is also testing that the `value(1)` doesn't make it to the expect listener at the end.

I kind of think it should. It makes the expectation clearer around what should happen when we have a circular dependency like this, so I think we should validate the entirety of what we expect to emit when a flatMapped stream ends "prematurely."

Reply:
That makes sense. I've swapped the code to what's in my example.